src/runner.ts

```diff
@@ -175,35 +175,50 @@ async function runLinter(options: Options, logger: Logger): Promise<LintResult>
 
 function resolveFilesAndProgram({ files, project, exclude, outputAbsolutePaths }: Options): { files: string[]; program?: ts.Program } {
     // remove single quotes which break matching on Windows when glob is passed in single quotes
-    const ignore = arrayify(exclude).map(trimSingleQuotes);
+    exclude = exclude.map(trimSingleQuotes);
 
     if (project === undefined) {
-        return { files: resolveGlobs(files, ignore, outputAbsolutePaths) };
+        return { files: resolveGlobs(files, exclude, outputAbsolutePaths) };
     }
 
     const projectPath = findTsconfig(project);
     if (projectPath === undefined) {
         throw new FatalError(`Invalid option for project: ${project}`);
     }
 
+    exclude = exclude.map((pattern) => path.resolve(pattern));
     const program = Linter.createProgram(projectPath);
     let filesFound: string[];
-    if (files === undefined || files.length === 0) {
-        filesFound = Linter.getFileNames(program);
-        if (ignore.length !== 0) {
-            const mm = ignore.map((pattern) => new Minimatch(path.resolve(pattern)));
-            filesFound = filesFound.filter((file) => !mm.some((matcher) => matcher.match(file)));
-        }
+    if (files.length === 0) {
+        filesFound = filterFiles(Linter.getFileNames(program), exclude, false);
     } else {
-        filesFound = resolveGlobs(files, ignore, outputAbsolutePaths);
+        files = files.map((f) => path.resolve(f));
+        filesFound = filterFiles(program.getSourceFiles().map((f) => f.fileName), files, true);
+        for (const file of files) {
+            if (!glob.hasMagic(file)) {
```

Comment:
Given a project that only includes `index.ts`.
If you run `tslint -p . '*.js' '*.ts'` you only lint `index.ts`. `*.js` doesn't match anything but is ignored because it's a glob pattern ("has magic").
On the other hand if you run `tslint -p . index.ts test/index.ts` you get an error because `test/index.ts` is not in the project. That's because it doesn't have magic. With the current implementation it doesn't even matter if the file exists.

Reply:
That makes sense; I meant to say -- can you leave a _code comment_